src/compiler.lisp

```diff
@@ -0,0 +1,660 @@
+;;;; compiler.lisp - functions for compiling .coal files
+
+(defpackage #:coalton-impl/compiler
+  (:use
+   #:cl)
+  (:local-nicknames
+   (#:codegen #:coalton-impl/codegen)
+   (#:entry #:coalton-impl/entry)
+   (#:error #:coalton-impl/error)
+   (#:parser #:coalton-impl/parser)
+   (#:settings #:coalton-impl/settings)
+   (#:tc #:coalton-impl/typechecker)
+   (#:util #:coalton-impl/util))
+  (:export
+   #:compile-coalton
+   #:make-prologue))
+
+;;;; environment serialization
+
+;;; The code below defines a format for environment updates that can
+;;; be written as Common Lisp source. When Coalton code is embedded in
+;;; Lisp files using the coalton-toplevel macro, the arguments to the
+;;; environment mutators in the typechecker/environment package appear
+;;; in the macroexpansion as externalizable structures, in that the
+;;; make-load-form generic is defined for possible values, using
+;;; make-load-form-saving-slots.
+;;;
+;;; Producing Lisp source from Coalton requires a source-representable
```

Comment:
Isn't make-load-form source representable? You can write `#S(...)` in a regular lisp file.

Reply:
Is there a better way? A simple example, taking a kfun instance,

```
(defvar s (tc:make-kfun :from tc:+kstar+ :to tc:+kstar+))
```

is there a way to produce a (reasonably platform independent) source representation from 's'? make-load-form returns an allocator and initializer:

```
(multiple-value-list (make-load-form-saving-slots s))
  ==>
((SB-KERNEL::ALLOCATE-STRUCT 'COALTON-IMPL/TYPECHECKER/KINDS:KFUN)
 (PROGN
  (SETF (SB-KERNEL:%INSTANCE-REF * → * 1) '*)
  (SETF (SB-KERNEL:%INSTANCE-REF * → * 2) '*)))
```

Could you elaborate a little on how you're suggesting to use #s syntax?

Reply:
Why does it need to be platform independent?

Reply:
It doesn't have to be.

Reply:
In that case can we continue using make-load-form instead of defining bespoke serialization logic?